godot-core/src/builtin/array.rs

```diff
@@ -185,61 +282,189 @@ impl Array {
+        debug_assert!(!self.type_info().is_typed());
+        let type_info = TypeInfo::new::<T>();
+        if type_info.is_typed() {
+            let script = Variant::nil();
+            unsafe {
+                (interface_fn!(array_set_typed))(
+                    self.sys(),
+                    type_info.variant_type.sys(),
+                    type_info.class_name.string_sys(),
+                    script.var_sys(),
+                );
+            }
+        }
+    }
+}
+
+impl<T: VariantMetadata + FromVariant> TypedArray<T> {
+    /// Returns an iterator over the elements of the `TypedArray`. Note that this takes the array
+    /// by reference but returns its elements by value, since they are internally converted from
+    /// `Variant`.
+    ///
+    /// Notice that it's possible to modify the `TypedArray` through another reference while
+    /// iterating over it. This will not result in unsoundness or crashes, but will cause the
+    /// iterator to behave in an unspecified way.
+    pub fn iter_shared(&self) -> TypedArrayIterator<'_, T> {
+        TypedArrayIterator {
+            array: self,
+            next_idx: 0,
+        }
     }
 
-    /// Returns the value at the specified index as a `Variant`. To convert to a specific type, use
-    /// the available conversion methods on `Variant`, such as [`Variant::try_to`] or
-    /// [`Variant::to`].
+    /// Returns the value at the specified index.
     ///
     /// # Panics
     ///
     /// If `index` is out of bounds.
-    pub fn get(&self, index: usize) -> Variant {
+    pub fn get(&self, index: usize) -> T {
         let ptr = self.ptr(index);
-        // SAFETY: `ptr` just verified that the index is not out of bounds.
-        unsafe { (*ptr).clone() }
+        // SAFETY: `ptr()` just verified that the index is not out of bounds.
+        let variant = unsafe { &*ptr };
+        T::try_from_variant(variant).unwrap()
     }
 
     /// Returns the first element in the array, or `None` if the array is empty. Equivalent of
     /// `front()` in GDScript.
-    pub fn first(&self) -> Option<Variant> {
-        (!self.is_empty()).then(|| self.as_inner().front())
+    pub fn first(&self) -> Option<T> {
+        (!self.is_empty()).then(|| {
+            let variant = self.as_inner().front();
+            T::try_from_variant(&variant).unwrap()
+        })
     }
 
     /// Returns the last element in the array, or `None` if the array is empty. Equivalent of
     /// `back()` in GDScript.
-    pub fn last(&self) -> Option<Variant> {
-        (!self.is_empty()).then(|| self.as_inner().back())
+    pub fn last(&self) -> Option<T> {
+        (!self.is_empty()).then(|| {
+            let variant = self.as_inner().back();
+            T::try_from_variant(&variant).unwrap()
+        })
+    }
+
+    /// Returns the minimum value contained in the array if all elements are of comparable types.
+    /// If the elements can't be compared or the array is empty, `None` is returned.
```

Comment:
"If the elements can't be compared" -- when can this happen?

Reply:
Copied from upstream. At least ints vs. strings.